Documentation/proposals/202310-shard-autoscaling.md

```diff
@@ -0,0 +1,165 @@
+
+# Pitfalls of the current solution
+
+Prometheus can scale vertically pretty well, but instances that grow up to hundreds of GiB/TiB in WAL size start face several challenges:
+
+* Eventual cardinality explosions have a big blast radius if a single Prometheus instance is responsible for scraping the majority of monitored applications.
+* Recovering from a crash takes several minutes due to the WAL replay.
+
+While the Prometheus and PrometheusAgent CRDs implement sharding for scrape configurations, there are still pending issues and missing features that prevent adoption.
+
+* Environments where the load is undistributed throughout the day/week, operators need to constantly adjust number of shards or shard size.
+* Scaling down Prometheus Servers causes data loss, as there's no guarantee that stored data is persisted in centralized storage before the container is deleted.
+
+# Goals
+
+* Enable scaling of the Prometheus shards up and down via Horizontal Pod Autoscaler objects.
+
+# Non-Goals
+
+* Implement the autoscaling logic into the Prometheus operator.
+* Automatic balancing of targets to ensure homogeneous load across all Prometheus pods.
+
+# Audience
+
+* Users looking for a shard autoscaling mechanism with a good cost-reliability ratio.
+
+# How
+
+Today, there are a few strategies to measure the load of Prometheus instances.
+
+1. Perhaps you can use [Kubernetes Metrics API](https://kubernetes.io/docs/tasks/debug/debug-cluster/resource-metrics-pipeline/#metrics-api) to measure load in terms of CPU and/or Memory usage.
+
+2. Another strategy might be to measure the amount of samples ingested per second
+
+- Prometheus Agents: `rate(prometheus_agent_samples_appended_total[5m])`
+- Prometheus Servers: `rate(prometheus_tsdb_head_samples_appended_total[5m])`
+
+3. Or even measuring samples queried per second for Prometheus instances under heavy query load.
+
+- Prometheus Servers: `rate(prometheus_engine_query_samples_total[5m])`
+
+These are all data that could be configured as input for different Horizontal Pod Autoscalers, but how does HPAs knows if the amount of shard is bigger/lower than desired?
+
+## Scale subresource
+
+When working with any resource, including CRDs, HPAs depends on a subresource called [Scale](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#scale-subresource). This proposal suggests to implement the scale subresource for the Prometheus and PrometheusAgent CRDs. Instead of working on the "replicas" count, it will operate on the "shards" count because the purpose of scaling up (resp. down) is to distribute the same number of targets across more (resp. less) Prometheus instances.
```

Comment:
(nit)
```suggestion
When working with any resource, including CRDs, HPAs depend on a subresource called [Scale](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#scale-subresource). This proposal suggests to implement the scale subresource for the Prometheus and PrometheusAgent CRDs. Instead of working on the "replicas" count, it will operate on the "shards" count because the purpose of scaling up (resp. down) is to distribute the same number of targets across more (resp. less) Prometheus instances.
```